### PR TITLE
Remove usage of IndexSearcher#Search(Query, Collector) from monitor package

### PR DESCRIPTION
**Pull request by gsmiller**
Relates to: #12892

### CODE REVIEW

**Comment by gsmiller** [comment]
OK, I _think_ this is still a safe implementation but the change is that multiple collection threads will now be concurrently calling `CandidateMatcher#addMatch`. I believe this is safe but need another set of eyes on it. The reason I think this works is that we should never be interleaving calls with the same `doc` value since these are global docIDs.

**Comment by javanna** [comment]
I am not familiar with this code but I believe you are correct. This should work despite we use a plain `ArrayList`, because each thread should only get and compute its own docs. Testing would verify that, but I believe we never provide an executor to the searcher we use, neither in test nor in the prod code. Should we try and add that?

**Comment by javanna** [comment]
Also, thanks a lot for the help here, much appreciated!

**Review by romseygeek — APPROVED** [review]
LGTM, thanks @gsmiller 

Concurrency in the monitor is handled in Parallel/PartitionMatcher outside of the index searcher, and the searcher is always built internally and not exposed to clients so there's no opportunity for an executor to be set on it.

Separately, I wonder if it's worth adding some sugar methods to CollectorManager that wrap up single-threaded collectors?  I'm guessing that there are going to be many instances of the pattern "return a collector from newCollector and null from reduce", and it seems a bit of a retrograde step to replace one line of code with 13, especially if you don't actually use multithreading.

**Comment by gsmiller** [comment]
Thanks @romseygeek for providing more detail on the monitor implementation and confirming this is safe.

As for your suggestion of providing a sugar method in CollectorManager to do the wrapping for situations like this, I'm a bit conflicted. I took a crack at updating the PR here with a revision that does this for discussion. On one hand, I agree with you that it's pretty verbose (and annoying if I'm honest) for users to have to do their own Collector wrapping like this, so a convenience method would be nice. On the other hand, I'm concerned it could be trappy for users that don't read the "fine print" and end up creating concurrency bugs by incorrectly wrapping in a case where an IndexSearcher is using an Executor (and their Collector isn't thread-safe).

I think it's a balance of how often this single-threaded-collector pattern occurs and how potentially trappy this sugar method is. Thoughts?

**Comment by javanna** [comment]
There is a test that does the following:

```
QueryMatch.SIMPLE_MATCHER.createMatcher(searcher);
```

This uses public API and allows to provide a searcher externally, doesn't it? Do we need to enforce that the searcher used in monitor should not have an executor and/or should not try to apply concurrency?

File: lucene/monitor/src/java/org/apache/lucene/monitor/CollectingMatcher.java [hunk 1]
```diff
@@ -37,7 +38,7 @@ abstract class CollectingMatcher<T extends QueryMatch> extends CandidateMatcher<
   @Override
   public void matchQuery(final String queryId, Query matchQuery, Map<String, String> metadata)
       throws IOException {
-    searcher.search(matchQuery, new MatchCollector(queryId, scoreMode));
+    searcher.search(matchQuery, CollectorManager.wrap(new MatchCollector(queryId, scoreMode)));
```
Comment (javanna) [inline lucene/monitor/src/java/org/apache/lucene/monitor/CollectingMatcher.java:41]: I left a comment about whether this is entirely safe, I was under the impression that one can provide an external searcher but I am not familiar with monitor code ;)

Comment (romseygeek) [reply lucene/monitor/src/java/org/apache/lucene/monitor/CollectingMatcher.java:41]: I guess you *could* call it with an externally built searcher, but it wouldn't be very useful as the whole point is that it's a searcher over a batch of documents submitted to the monitor for checking.  It's public to allow clients to build wrapping matchers.

There may be a nicer way to do this though, let me experiment.

Comment (romseygeek) [reply lucene/monitor/src/java/org/apache/lucene/monitor/CollectingMatcher.java:41]: I think we can make the public-facing methods all take `DocumentBatch` here rather than `IndexSearcher`, which will make building the index searcher an implementation detail.  I'll put together a PR.

File: lucene/core/src/java/org/apache/lucene/search/CollectorManager.java [hunk 2]
```diff
@@ -46,4 +48,28 @@ public interface CollectorManager<C extends Collector, T> {
...
+   *
+   * <p>Note: This is only safe to use when {@code IndexSearcher} is created with no executor (see:
+   * {@link IndexSearcher#IndexSearcher(IndexReader, Executor)}), or the provided collector is
+   * threadsafe.
+   */
+  static <C extends Collector> CollectorManager<C, ?> wrap(C in) {
+    return new CollectorManager<C, Void>() {
+      @Override
+      public C newCollector() {
+        return in;
```
[the remark this answers is not in the document]
Comment (gsmiller) [reply lucene/core/src/java/org/apache/lucene/search/CollectorManager.java:66]: I think `IndexSearcher` is going to [call this multiple times](https://github.com/apache/lucene/blob/c21bc5405be258bcebf6b93ceb13f9554c214b2c/lucene/core/src/java/org/apache/lucene/search/IndexSearcher.java#L652) (once per leaf slice) even when not doing multi-threaded search right? So I'm not sure we can do this?

[the remark this answers is not in the document]
Comment (javanna) [reply lucene/core/src/java/org/apache/lucene/search/CollectorManager.java:66]: not if an executor is not provided, which I thought was why this was safe to do. In that case we create a single slice.

File: lucene/core/src/java/org/apache/lucene/search/CollectorManager.java [hunk 3]
```diff
@@ -46,4 +48,28 @@ public interface CollectorManager<C extends Collector, T> {
...
+      }
+
+      @Override
+      public Void reduce(Collection<C> collectors) {
+        return null;
```
[the remark this answers is not in the document]
Comment (gsmiller) [reply lucene/core/src/java/org/apache/lucene/search/CollectorManager.java:71]: I might lean towards `IllegalStateException` for `newCollector` but a simple `assert` here (since the user really shouldn't be able to get into a state where `newCollector` was only called once but `reduce` was passed a list with more than one collector). WDYT?

File: lucene/core/src/java/org/apache/lucene/search/CollectorManager.java [hunk 4]
```diff
@@ -46,4 +48,28 @@ public interface CollectorManager<C extends Collector, T> {
...
+   * Wrap a provided {@link Collector} with a thin {@code CollectorManager} wrapper for use with
+   * {@link IndexSearcher#search(Query, CollectorManager)}. The wrapping {@code CollectorManager}
+   * provides no {@link CollectorManager#reduce(Collection)} implementation, so the wrapped {@code
+   * Collector} needs to do all relevant work while collecting.
+   *
+   * <p>Note: This is only safe to use when {@code IndexSearcher} is created with no executor (see:
+   * {@link IndexSearcher#IndexSearcher(IndexReader, Executor)}), or the provided collector is
+   * threadsafe.
+   */
+  static <C extends Collector> CollectorManager<C, ?> wrap(C in) {
```
[the remark this answers is not in the document]
Comment (gsmiller) [reply lucene/core/src/java/org/apache/lucene/search/CollectorManager.java:62]: Makes sense. I'll revise the name (but if you have suggestions, I'm open to them; naming can be tricky)

File: lucene/core/src/java/org/apache/lucene/search/CollectorManager.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (msfroh) [reply lucene/core/src/java/org/apache/lucene/search/CollectorManager.java:71]: Can we assume that the `Collector`s passed to `reduce` were produced by calling `newCollector` on this collector?

If so, the assertion that `newCollector` was called at most once should be sufficient.

[the remark this answers is not in the document]
Comment (javanna) [reply lucene/core/src/java/org/apache/lucene/search/CollectorManager.java:71]: Right, agreed. We can assume that the collection holds the collectors that newCollector returned.

**Comment by gsmiller** [comment]
OK, I think the latest iteration of this is ready to merge, pending any remaining feedback, concerns, etc. I voiced some concerns earlier about the static wrapping method, wanting to make sure we don't provide some trappy implementation for users, but the suggestions around error checking/asserts convinced me this is reasonable (still don't love it, but I think it's reasonable right now).

I'll get this merged in the next couple days if I don't hear any additional concerns, otherwise happy to iterate.

**Review by javanna — APPROVED** [review]
ship it! :) thanks again for picking this up. I think we will reuse the method you added in a couple of other places.

**Comment by javanna** [comment]
Heads up: I saw there were merge conflicts caused by my recent push hence I went ahead and addressed those.

**Comment by gsmiller** [comment]
Thanks @javanna, @romseygeek !  FYI @msfroh just merged and back ported. I saw you had another PR waiting on this.

**Comment by javanna** [comment]
I've been changing my mind a few times on the topic, but I am concluding that it probably was not a good idea to expose the sequential collector manager. I think we are ok using it internally at our own risk, for situations where we know for sure that we can't possibly have a searcher with an executor set. But exposing it publicly makes it too easy for users to get unexpected failures, and the disclaimer "don't use this with a searcher that has an executor set" is kind of odd. What if those users do have other queries that make already use of concurrency, and use the same searcher for these others that get converted to leveraging the sequential collector manager? It is also trappy that they get errors only once there's more than one slice, so the behaviour may be hard to follow for users.

I would propose that we make the new collector manager private with a bug disclaimer on when it should be used, and make a plan to remove it perhaps in the medium to long term. We should really design all new features with concurrency in mind, and migrate old ones to support concurrency.

There is some urgency around this especially for the 9.12 release, I am happy to open a PR.
